### PR TITLE
Allow DataParallel returning dict and any other instances of iterable custom classes

### PR DESCRIPTION
**Pull request by dalegebit**
Returning a dict or instances of custom classes can be very helpful, especially for complex networks. For exmaple, the RNN state of a complex RNN model is usually wrapped inside a custom class.

### CODE REVIEW

File: torch/nn/parallel/scatter_gather.py [hunk 1]
```diff
@@ -46,5 +46,8 @@ def gather_map(outputs):
             return Gather.apply(target_device, dim, *outputs)
         if out is None:
             return None
-        return type(out)(map(gather_map, zip(*outputs)))
+        if isinstance(out, dict):
+            return dict(zip(out.keys(),
+                            map(gather_map, zip(*[d.values() for d in outputs]))))
```
Comment (SsnL) [inline torch/nn/parallel/scatter_gather.py]: Correct me if I'm wrong, but I don't think that `keys()` is necessarily ordered in the same way as `values()`. Could you do instead 
```python
keys = tuple(out.keys())
return dict([(k, gather_map([d[k] for d in outputs])) for k in keys])
```

Comment (apaszke) [reply torch/nn/parallel/scatter_gather.py]: or better yet, iterate over `.items()`

Comment (SsnL) [reply torch/nn/parallel/scatter_gather.py]: @apaszke Not necessarily. It is not guaranteed that every dict's `.items()` is of the same order. So it's still better to index into every dict using a reference dict's keys.

Comment (apaszke) [reply torch/nn/parallel/scatter_gather.py]: Ah I haven't noticed that these are two different dicts. You're right.

**Comment by weiyangfb** [comment]
closing this because similar PR is already merged: https://github.com/pytorch/pytorch/pull/6113

**Pull request closed without merging**